File: widgets/symphony_summary/src/widget.ts
// For licensing see accompanying LICENSE file.
// Copyright (C) 2023 Apple Inc. All Rights Reserved.

import { MODULE_NAME, MODULE_VERSION } from './version';
import {
  DOMWidgetModel,
  DOMWidgetView,
  ISerializers,
} from '@jupyter-widgets/base';

import {serializers, JupyterWidget } from '@apple/symphony-lib';

import SymphonySummary from './SymphonySummary.svelte';

export class SymphonySummaryModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: SymphonySummaryModel.model_name,
      _model_module: SymphonySummaryModel.model_module,
      _model_module_version: SymphonySummaryModel.model_module_version,
      _view_name: SymphonySummaryModel.view_name,
      _view_module: SymphonySummaryModel.view_module,
      _view_module_version: SymphonySummaryModel.view_module_version,
    };
  }
  static serializers: ISerializers = serializers;

  static model_name = 'SymphonySummaryModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'SymphonySummaryView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class SymphonySummaryView extends DOMWidgetView {
  render() {
    new JupyterWidget({
      target: this.el,
      props: {
        model: this.model,
        widget: SymphonySummary,
      },
    });
  }
}

File: widgets/symphony_summary/symphony_summary/widget.py
# For licensing see accompanying LICENSE file.
# Copyright (C) 2023 Apple Inc. All Rights Reserved.

from pathlib import Path
from typing import List
from ._frontend import module_name, module_version

import symphony_ui
from traitlets import Unicode


class SymphonySummary(symphony_ui.SymphonyWidget):
    _model_name = Unicode('SymphonySummaryModel').tag(sync=True)
    _view_name = Unicode('SymphonySummaryView').tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)
    name = 'SymphonySummary'
    description = "A Symphony component that visualizes an overview of a dataset"

    def __init__(self,
                 width: str = 'XXL',
                 height: str = 'M',
                 page: str = 'Summary',
                 summary_elements: List[symphony_ui.SummaryElement] = [],
                 **kwargs):
        """A Symphony component that visualizes an overview of a dataset

        Parameters
        ----------
        width : str, optional
            By default "XXL".
        height : str, optional
            By default "M".
        page : str, optional
            Which page of the dashboard to show the widget on, by default "Familiarity"
        summary_elements : List[SummaryElement], optional
            Vega charts or metrics to display, by default []

        Returns
        -------
        SymphonySummary
        """
        super().__init__(**kwargs)
        self.width: str = width
        self.height: str = height
        self.page: str = page
        self.summary_elements: List[symphony_ui.SummaryElement] = summary_elements
        self.spec = symphony_ui.SymphonySummarySpec(
            width=width,
            height=height,
            page=page,
            name=self.name,
            description=self.description,
            summary_elements=summary_elements,
        )

    def js_path(self):
        app_files_path = Path(
            (Path(__file__).parent), 'standalone', 'widgets')
        js_path = Path(app_files_path, 'SymphonySummary.js')
        map_path = Path(app_files_path, 'SymphonySummary.js.map')
        return js_path, map_path
